File: 454-4sum-ii/454-4sum-ii.py
from collections import defaultdict
class Solution:
    def fourSumCount(self, nums1: List[int], nums2: List[int], nums3: List[int], nums4: List[int]) -> int:
        hmap = defaultdict(int)
        for i in nums1:
            for j in nums2:
                hmap[i+j] += 1
        ans = 0
        for i in nums3:
            for j in nums4:
                add = -(i+j)
                ans += hmap[add]
        return ans

File: 454-4sum-ii/454-4sum-ii.java
class Solution {
    public int fourSumCount(int[] nums1, int[] nums2, int[] nums3, int[] nums4) {
        Map<Integer, Integer> hmap = new HashMap<>();
        for(int i: nums1) {
            for(int j: nums2) {
                int sum = i+j;
                if (hmap.containsKey(i+j)) {
                    hmap.put((i+j), hmap.get(i+j)+1);
                }
                else {
                    hmap.put((i+j), 1);
                }
            }
        }
        int c = 0;
        for(int i: nums3) {
            for(int j: nums4) {
                int negative = -(i+j);
                if (hmap.containsKey(negative)) {
                    c += hmap.get(negative);
                }
                    
            }
        }
        return c;
    }
}